--- a/3rdparty/cmake/FindPytorch.cmake
+++ b/3rdparty/cmake/FindPytorch.cmake
@@ -17,9 +17,9 @@
 macro(translate_arch_string input output)
     if("${input}" MATCHES "[0-9]+-real")
-        string(REGEX REPLACE "([1-9])([0-9])-real" "\\1.\\2" version "${input}")
+        string(REGEX REPLACE "([1-9][0-9]?)([0-9])-real" "\\1.\\2" version "${input}")
     elseif("${input}" MATCHES "([0-9]+)")
-        string(REGEX REPLACE "([1-9])([0-9])" "\\1.\\2+PTX" version "${input}")
+        string(REGEX REPLACE "([1-9][0-9]?)([0-9])" "\\1.\\2+PTX" version "${input}")
     elseif("${input}" STREQUAL "native")
         set(version "Auto")
     else()
